File: Android/app/src/main/java/ru/inwords/inwords/data/repository/translation/TranslationWordsCacheRepository.java
package ru.inwords.inwords.data.repository.translation;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.subjects.BehaviorSubject;
import io.reactivex.subjects.Subject;
import ru.inwords.inwords.core.util.SchedulersFacade;
import ru.inwords.inwords.data.dto.WordTranslation;

public class TranslationWordsCacheRepository implements TranslationWordsLocalRepository {
    private final ConcurrentHashMap<WordTranslation, WordTranslation> list;

    private final Subject<List<WordTranslation>> behaviorSubject;

    @Inject
    TranslationWordsCacheRepository() {
        this.list = new ConcurrentHashMap<>();

        behaviorSubject = BehaviorSubject.createDefault(Collections.list(list.elements()));
    }

    @Override
    public Observable<WordTranslation> getTranslation(String word) {
        return Observable.fromIterable(list.values())
                .filter(wordTranslation -> wordTranslation.getWordForeign().equals(word) ||
                        wordTranslation.getWordNative().equals(word));
    }

    @Override
    public Observable<WordTranslation> getByOne() {
        return getList()
                .flatMap(Observable::fromIterable);
    }

    @Override
    public Observable<List<WordTranslation>> getList() {
        return behaviorSubject
                .subscribeOn(SchedulersFacade.io())
                //.filter(wordTranslations -> !wordTranslations.isEmpty())
                .map(ArrayList::new);
    }

    @Override
    public Single<WordTranslation> addReplace(WordTranslation wordTranslation) {
        return Single.defer(() -> {
            WordTranslation storedItem = list.get(wordTranslation);
            WordTranslation itemToAdd;
            if (storedItem == null){
                itemToAdd = new WordTranslation(wordTranslation.getWordForeign(), wordTranslation.getWordNative());
            } else {
                itemToAdd = wordTranslation;
            }
            list.put(itemToAdd, itemToAdd);

            return Single.just(itemToAdd)
                    .doOnSuccess((o) -> publish());
        });
    }

    @Override
    public Single<List<WordTranslation>> addReplaceAll(List<WordTranslation> wordTranslations) {
        if (wordTranslations.isEmpty()){
            return Single.just(wordTranslations);
        }

        return Single.defer(() -> {
            //list.removeAll(wordTranslations);
            for (WordTranslation wordTranslation : wordTranslations) {
                list.put(wordTranslation, wordTranslation);
            }

            return Single.just(wordTranslations)
                    .doOnSuccess((o) -> publish());
        });
    }

    @Override
    public Completable removeAll(List<WordTranslation> wordTranslations) {
        if (wordTranslations.isEmpty()){
            return Completable.complete();
        }

        return Completable.fromCallable(() -> {
            list.keySet().removeAll(wordTranslations);

            publish();

            return true;
        });
    }

    @Override
    public Completable removeAllServerIds(List<Integer> serverIds) {
        if (serverIds.isEmpty()){
            return Completable.complete();
        }

        return Completable.fromCallable(() -> {
            for (Integer serverId : serverIds) {
                for (Iterator<WordTranslation> it = list.values().iterator(); it.hasNext(); ) {
                    WordTranslation next = it.next();
                    if (next.getServerId() == serverId) {
                        it.remove();
                    }
                }
            }

            publish();

            return true;
        });
    }

    private void publish() {
        behaviorSubject.onNext(Collections.list(list.elements()));
    }
}

File: Android/app/src/main/java/ru/inwords/inwords/data/repository/translation/TranslationWordsCacheRepository.kt
package ru.inwords.inwords.data.repository.translation

import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.subjects.BehaviorSubject
import ru.inwords.inwords.core.util.SchedulersFacade
import ru.inwords.inwords.data.dto.WordTranslation
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject

class TranslationWordsCacheRepository @Inject
internal constructor() : TranslationWordsLocalRepository {
    private val list: ConcurrentHashMap<WordTranslation, WordTranslation> = ConcurrentHashMap()

    private val behaviorSubject = BehaviorSubject.createDefault(Collections.list(list.elements()))

    override fun getByOne(): Observable<WordTranslation> {
        return getList()
                .flatMap { Observable.fromIterable(it) }
    }

    override fun getList(): Observable<List<WordTranslation>> {
        return behaviorSubject
                .observeOn(SchedulersFacade.io())
                //.filter(wordTranslations -> !wordTranslations.isEmpty())
                .map { ArrayList(it) }
    }

    override fun addReplace(wordTranslation: WordTranslation): Single<WordTranslation> {
        return Single.defer {
            val storedItem = list[wordTranslation]
            val itemToAdd = if (storedItem == null) {
                WordTranslation(wordTranslation.wordForeign, wordTranslation.wordNative)
            } else {
                wordTranslation
            }
            list[itemToAdd] = itemToAdd

            Single.just(itemToAdd)
                    .doOnSuccess { publish() }
        }
    }

    override fun addReplaceAll(wordTranslations: List<WordTranslation>): Single<List<WordTranslation>> {
        return if (wordTranslations.isEmpty()) {
            Single.just(wordTranslations)
        } else Single.defer {
            //list.removeAll(wordTranslations);
            for (wordTranslation in wordTranslations) {
                list[wordTranslation] = wordTranslation
            }

            Single.just(wordTranslations)
                    .doOnSuccess { publish() }
        }

    }

    override fun removeAll(wordTranslations: List<WordTranslation>): Completable {
        return if (wordTranslations.isEmpty()) {
            Completable.complete()
        } else {
            Completable.fromCallable {
                list.keys.removeAll(wordTranslations)

                publish()

                true
            }
        }

    }

    override fun removeAllServerIds(serverIds: List<Int>): Completable {
        return if (serverIds.isEmpty()) {
            Completable.complete()
        } else {
            Completable.fromCallable {
                for (serverId in serverIds) {
                    val it = list.values.iterator()
                    while (it.hasNext()) {
                        val next = it.next()
                        if (next.serverId == serverId) {
                            it.remove()
                        }
                    }
                }

                publish()

                true
            }
        }
    }

    override fun clear() = list.clear()

    private fun publish() {
        behaviorSubject.onNext(Collections.list(list.elements()))
    }
}
